draft-ietf-quic-tls.md

```diff
@@ -318,25 +318,24 @@ long as they are associated with the same encryption level. For instance, an
 implementation might bundle a Handshake message and an ACK for some Handshake
 data into the same packet.
 
-Each encryption level has a specific list of frames which may appear in it. The
-rules here generalize those of TLS, in that frames associated with establishing
-the connection can usually appear at any encryption level, whereas those
-associated with transferring data can only appear in the 0-RTT and 1-RTT
+Some frames are prohibited in different encryption levels, others cannot be
+sent. The rules here generalize those of TLS, in that frames associated with
+establishing the connection can usually appear at any encryption level, whereas
+those associated with transferring data can only appear in the 0-RTT and 1-RTT
 encryption levels:
 
-- CRYPTO frames MAY appear in packets of any encryption level except 0-RTT.
-
-- CONNECTION_CLOSE MAY appear in packets of any encryption level other than
-  0-RTT.
-
 - PADDING frames MAY appear in packets of any encryption level.
 
+- CRYPTO and CONNECTION_CLOSE frames MAY appear in packets of any encryption
```

Comment:
You can't really prevent implementations from sending CONNECTION_CLOSE in 0-RTT. How would the peer punish them for misbehaving, by closing the connection? So we might as well allow it and read the error code and error message.

Reply:
The fear I have in allowing the CONNECTION_CLOSE frame to be sent in 0-RTT packet is that it might cause confusion among the people reading the specification.

Currently, an endpoint is expected to send a CONNECTION_CLOSE frame using a packet that it knows that the peer can decrypt (see https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.section.10.3). I think we should retain that principle, and based on that principle, an endpoint would never be sending a CONNECTION_CLOSE frame in a 0-RTT packet.

Reply:
On CONNECTION_CLOSE: There are cases why you might send possible not received messages with minimal overhead and close immediately. Like a sensor reading. This isn't really idempotent, but it might be good enough for the application. MQTT added the MQTT-SN variant that accepts single packet messages on UDP although technically the connection is still open.

Reply:
But there are other problems: I suppose you can't send multiple such messages without a new 0-RTT token. Maybe a use case for v2?

Reply:
There is no reason to send CONNECTION_CLOSE in 0-RTT.  CONNECTION_CLOSE has those rules (which need to be clarified, but they won't materially change), which I'd say are fine.  During the handshake, implementations have to be specially aware of where CONNECTION_CLOSE lands, that's all.